testing/base/source_gen.cpp

```diff
@@ -0,0 +1,676 @@
+  id_length_counts[5] = 20;
+  // 7 characters   [2380]  ███████████████████
+  id_length_counts[6] = 20;
+  // 8 characters   [1786]  ██████████████▎
+  id_length_counts[7] = 18;
+  // 9 characters   [1397]  ███████████▏
+  id_length_counts[8] = 12;
+  // 10 characters  [ 739]  █████▉
+  id_length_counts[9] = 12;
+  // 11 characters  [ 779]  ██████▎
+  id_length_counts[10] = 12;
+  // 12 characters  [1344]  ██████████▊
+  id_length_counts[11] = 12;
+  // 13 characters  [ 498]  ████
+  id_length_counts[12] = 5;
+  // 14 characters  [ 284]  ██▎
+  id_length_counts[13] = 3;
+  // 15 characters  [ 172]  █▍
+  // 16 characters  [ 278]  ██▎
+  // 17 characters  [ 191]  █▌
+  // 18 characters  [ 207]  █▋
+  for (int i = 14; i < 18; ++i) {
+    id_length_counts[i] = 2;
+  }
+  // 19 - 63 characters are all <100 but non-zero, and we map them to 1 for
+  // coverage despite slightly over weighting the tail.
+  for (int i = 18; i < 64; ++i) {
+    id_length_counts[i] = 1;
+  }
+  return id_length_counts;
+}();
+
+// A helper to sum elements of a range.
+template <typename T>
+static auto Sum(const T& range) -> int {
+  return std::accumulate(range.begin(), range.end(), 0);
+}
+
+// A template function that implements the common logic of `GetIds` and
+// `GetUniqueIds`. Most parameters correspond to the parameters of those
+// functions. Additionally, an `AppendIds` callable is provided to implement the
+// appending operation.
+//
+// The main functionality provided here is collecting the correct number of
+// identifiers from each of the lengths in the range [min_length, max_length]
+// and either in our default representative distribution or a uniform
+// distribution.
+//
+// Note that this template must be defined prior to its use below.
+template <typename AppendIds>
+auto SourceGen::GetIdsImpl(int number, int min_length, int max_length,
+                           bool uniform, AppendIds append_ids)
+    -> llvm::SmallVector<llvm::StringRef> {
+  CARBON_CHECK(min_length <= max_length);
+  CARBON_CHECK(uniform || max_length <= 64)
+      << "Cannot produce a meaningful non-uniform distribution of lengths "
+         "longer than 64 as those are exceedingly rare in our observed data "
+         "sets.";
+
+  llvm::SmallVector<llvm::StringRef> ids;
+  ids.reserve(number);
+
+  // First, compute how many identifiers of each size we'll need.
+  int count_sum =
+      uniform ? (max_length - min_length) + 1
+              : Sum(llvm::ArrayRef(IdLengthCounts)
+                        .slice(min_length - 1, max_length - min_length + 1));
+  CARBON_CHECK(count_sum >= 1);
+  int number_rem = number % count_sum;
+  for (int length : llvm::seq_inclusive(min_length, max_length)) {
+    // Scale this length if non-uniform.
+    int scale = uniform ? 1 : IdLengthCounts[length - 1];
+    int length_count = (number / count_sum) * scale;
+    if (number_rem > 0) {
+      length_count += std::min(scale, number_rem);
+      number_rem -= scale;
+    }
```

Comment:
For small counts, this'll overallocate towards the bottom end. Eg, if I ask for 100 identifiers with lengths 4 to 64, I'll get {50, 20, 20, 10, 0, 0, 0, ...}. How about:
```suggestion
    int length_count = (number * scale) / count_sum;
    if (number_rem > 0 && length_count * count_sum < number * scale) {
      ++length_count;
      --number_rem;
    }
```
That is: get the floor of the ideal count (which will be off by less than 1) and then add one to the first `number_rem` buckets that we rounded down.

For my example, I think that'd give something more like {21, 9, 9, 9, 8, 5, 5, 5, 5, 3, 2, <19 1s>}, which seems a bit more representative. It still skews a little towards shorter identifiers, but that's probably not that unreasonable.

Reply:
I tried this, and the math doesn't seem to check out.

Specifically, we don't actually get the right number of identifiers with the code as suggested (the check below fails). I've tried several things, and none of them seem to reliably produce what you're suggesting here, but probably just haven't hit on the right model.

I feel like for this to work we'd need to introduce some kind of *clamping* of the distribution. I'm not opposed, but I don't have the math for it, and it seems like this could be a follow-up if you're interested? I don't really have a strong use case for making the small number of identifiers that much more representative TBH, so unclear this is worthwhile.